sale_by_packaging/models/sale_order_line.py

```diff
@@ -0,0 +1,160 @@
+                        "This product packaging must be set as 'Can be sold' in"
+                        " order to be used on a sale order."
+                    ),
+                },
+            }
+        return super()._onchange_product_packaging()
+
+    @api.constrains("product_id", "product_packaging")
+    def _check_product_packaging_sell_only_by_packaging(self):
+        for line in self:
+            if line.product_id.sell_only_by_packaging and not line.product_packaging:
+                raise ValidationError(
+                    _(
+                        "Product %s can only be sold with a packaging."
+                        % line.product_id.name
+                    )
+                )
+
+    @api.onchange("product_id")
+    def product_id_change(self):
+        res = super().product_id_change()
+        if self.product_id.sell_only_by_packaging:
+            first_packaging = fields.first(self.product_id.packaging_ids)
+            self.update(
+                {
+                    "product_packaging": first_packaging.id,
+                    "product_uom_qty": first_packaging.qty,
+                }
+            )
+        return res
+
+    @api.onchange("product_uom_qty")
+    def _onchange_product_uom_qty(self):
+        res = super()._onchange_product_uom_qty()
+        if not res:
+            res = self._check_qty_is_pack_multiple()
+        return res
+
+    def _check_qty_is_pack_multiple(self):
+        """ Check only for product with sell_only_by_packaging
+        """
+        # and we dont want to have this warning when we had the product
+        if self.product_id.sell_only_by_packaging:
+            if not self._get_product_packaging_having_multiple_qty(
+                self.product_id, self.product_uom_qty, self.product_uom
+            ):
+                warning_msg = {
+                    "title": _("Product quantity can not be packed"),
+                    "message": _(
+                        "For the product {prod}\n"
+                        "The {qty} is not the multiple of any pack.\n"
+                        "Please add a package"
+                    ).format(prod=self.product_id.name, qty=self.product_uom_qty),
+                }
+                return {"warning": warning_msg}
+        return {}
+
+    def _get_product_packaging_having_multiple_qty(self, product, qty, uom):
+        if uom != product.uom_id:
+            qty = uom._compute_quantity(qty, product.uom_id)
+        return product.get_first_packaging_with_multiple_qty(qty)
+
+    def write(self, vals):
+        """Auto assign packaging if needed"""
+        fields_to_check = ["product_id", "product_uom_qty", "product_uom"]
+        if vals.get("product_packaging") or not any(
+            fname in vals for fname in fields_to_check
+        ):
+            return super().write(vals)
+        for line in self:
+            line_vals = vals.copy()
+            line_vals.update(self._write_auto_assign_packaging(line_vals))
```

Comment:
shouldn't be called on line?
```suggestion
            line_vals.update(line._write_auto_assign_packaging(line_vals))
```